File: yt/yt/flow/examples/kotlin/static_table_join/static_table_join/src/main/kotlin/tech/ytsaurus/flow/examples/statictablejoin/ReferenceLoader.kt
package tech.ytsaurus.flow.examples.statictablejoin

import tech.ytsaurus.flow.computation.OutputCollector
import tech.ytsaurus.flow.context.RuntimeContext
import tech.ytsaurus.flow.function.RowFunction
import tech.ytsaurus.flow.row.ExtendedMessage
import tech.ytsaurus.flow.spring.FlowComputation
import tech.ytsaurus.flow.state.ExternalStateDescriptor
import tech.ytsaurus.flow.state.StateDescriptors

// [BEGIN registration]
@FlowComputation(id = "reference_loader")
// [END registration]

/**
 * Trims and lowercases the reference name, storing it in keyed external state.
 */
open class ReferenceLoader : RowFunction {
    companion object {
        private val REFERENCE_STATE: ExternalStateDescriptor =
            StateDescriptors.external("/reference_state")
    }

    // [BEGIN on_message]
    override fun onMessage(message: ExtendedMessage, output: OutputCollector, ctx: RuntimeContext) {
        val normalizedName = message.get("name", String::class.java)?.trim()?.lowercase() ?: return

        val stateAccessor = ctx.getState(REFERENCE_STATE, message)
        stateAccessor.set(
            stateAccessor.orDefault.toBuilder()
                .set("normalized_name", normalizedName)
                .finish()
        )
    }
    // [END on_message]
}

File: yt/yt/flow/examples/go/static_table_join/reference_loader.go
package main

import (
	"context"
	"strings"

	"go.ytsaurus.tech/yt/go/flow"
)

type referenceState struct {
	NormalizedName *string `yson:"normalized_name"`
}

type referenceLoader struct{}

var _ flow.RowFunction = (*referenceLoader)(nil)

func (*referenceLoader) OnMessage(
	ctx context.Context,
	rt flow.Runtime,
	msg flow.ExtendedMessage,
	out flow.OutputCollector,
) error {
	var reference referenceMessage
	if err := msg.ConvertTo(&reference); err != nil {
		return err
	}

	state, err := flow.OpenExternalState(rt, referenceStateName, msg)
	if err != nil {
		return err
	}

	normalized := normalize(reference.Name)
	return state.ConvertFrom(&referenceState{NormalizedName: &normalized})
}

func normalize(name string) string {
	return strings.ToLower(strings.TrimSpace(name))
}
